controllers/components/kserve/kserve_support.go

```diff
@@ -0,0 +1,300 @@
+	cluster.SelfManagedRhods: {odhModelControllerComponentName},
+}
+
+var clusterRolePredicate = predicate.Funcs{
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		notAllowedNames := []string{"kserve-admin"}
+		for _, notallowedName := range notAllowedNames {
+			if e.ObjectNew.GetName() == notallowedName {
+				return false
+			}
+		}
+		return true
+	},
+}
+
+var compareHashPredicate = predicate.Funcs{
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		oldUnstructured, err := resources.ToUnstructured(e.ObjectOld.DeepCopyObject())
+		if err != nil {
+			return true
+		}
+		newUnstructured, err := resources.ToUnstructured(e.ObjectNew.DeepCopyObject())
+		if err != nil {
+			return true
+		}
+
+		oldHash, err := resources.Hash(oldUnstructured)
+		if err != nil {
+			return true
+		}
+		newHash, err := resources.Hash(newUnstructured)
+		if err != nil {
+			return true
+		}
+
+		return base64.RawURLEncoding.EncodeToString(oldHash) != base64.RawURLEncoding.EncodeToString(newHash)
+	},
+}
+
+func kserveManifestInfo(sourcePath string) odhtypes.ManifestInfo {
+	return odhtypes.ManifestInfo{
+		Path:       deploy.DefaultManifestPath,
+		ContextDir: componentName,
+		SourcePath: sourcePath,
+	}
+}
+
+func odhModelControllerManifestInfo(sourcePath string) odhtypes.ManifestInfo {
+	return odhtypes.ManifestInfo{
+		Path:       deploy.DefaultManifestPath,
+		ContextDir: odhModelControllerComponentName,
+		SourcePath: sourcePath,
+	}
+}
+
+func checkDependentOperators(ctx context.Context, cli client.Client) *multierror.Error {
+	var multiErr *multierror.Error
+
+	if found, err := cluster.OperatorExists(ctx, cli, serviceMeshOperator); err != nil {
+		multiErr = multierror.Append(multiErr, err)
+	} else if !found {
+		err = fmt.Errorf("operator %s not found. Please install the operator before enabling %s component",
+			serviceMeshOperator, componentName)
+		multiErr = multierror.Append(multiErr, err)
+	}
+
+	if found, err := cluster.OperatorExists(ctx, cli, serverlessOperator); err != nil {
+		multiErr = multierror.Append(multiErr, err)
+	} else if !found {
+		err = fmt.Errorf("operator %s not found. Please install the operator before enabling %s component",
+			serverlessOperator, componentName)
+		multiErr = multierror.Append(multiErr, err)
+	}
+	return multiErr
+}
+
+func configureServerlessFeatures(dsciSpec *dsciv1.DSCInitializationSpec, kserve *componentsv1.Kserve) feature.FeaturesProvider {
+	return func(registry feature.FeaturesRegistry) error {
+		servingDeployment := feature.Define("serverless-serving-deployment").
+			Manifests(
+				manifest.Location(Resources.Location).
+					Include(
+						path.Join(Resources.InstallDir),
+					),
+			).
+			WithData(
+				serverless.FeatureData.IngressDomain.Define(&kserve.Spec.Serving).AsAction(),
+				serverless.FeatureData.Serving.Define(&kserve.Spec.Serving).AsAction(),
+				servicemesh.FeatureData.ControlPlane.Define(dsciSpec).AsAction(),
+			).
+			PreConditions(
```

Comment:
not sure how `checkDependentOperators` and those preconditions are actually working together

Reply:
I've left these PreConditions for now, but I think they could be revisited if we decide to pull the use of FeatureTrackers from this kserve reconciler.

There are 4 preconditions here, two of which overlap with the checkDependentOperators (which I've replaced with an action to check at the start, which I think is what you were suggesting). The two that overlap are to check the existence of the ServiceMesh and Serverless operators. They check in slightly different ways though: one looks at `OperatorConditions`, where the other looks at `Subscriptions`...I'm not sure if it's necessary to check both. It boils down to either `SubscriptionExists()` or `OperatorExists()`, both of which are functions defined here: https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/pkg/cluster/operator.go